### DISCUSSION TITLE
Have there been changes to the way Jacktrip installs Jack?  We're seeing weird connection/Jack issues

### DISCUSSION THREAD

**Discussion by OConnorStP** [jacktrip/jacktrip#941] [General]
Hi all,

Over the past few months, the process of starting people's Jacktrip session has gotten weirdly harder.  All kinds of quirky issues like one-way connections (reminding me of the old firewall port-forwarding days), Qjackctl failures etc.  In many cases we've reverted to starting Jack from the command line, but even that is getting bumpy with Jack failing because of permission problem in a recent session with a player.

In earlier times, I'd have people install Jack and then install Jacktrip.  In more recent time, Jacktrip does it.  Which leads me to this thread -- I'm wondering if that code got jostled in all the other changes that have happening.  Might it be better to separate the Jack install again, instead of having Jacktrip do it?

**Comment by psiborg112** [comment]
JackTrip shouldn't install JACK. (I know my binaries don't and just taking a look at the official releases, there's absolutely nothing to suggest they do either.) If JACK isn't installed, JackTrip will fallback to using RtAudio. (Are there old installs of JACK that are hanging around on some of these machines?)

**Reply by SynthiaPayne** [reply]
The latest issue on the jack2 github is about the jack driver not uninstalling. 

https://github.com/jackaudio/jack2/issues/920#issuecomment-1445419483

**Comment by OConnorStP** [comment]
That's interesting!  I've got a session coming up where I can verify this, but I'm almost sure Jack rolls in with a Jacktrip install.  I'm pretty sure I've deleted Jack and Jacktrip (cleaned out all the /usr/local/bin/ stuff etc.) and then only run the Jacktrip installer and gone back to find that the Jack binaries are back in /usr/local/bin.  But leave this one on me to verify.  Should be later today.

**Comment by ntonnaett** [comment]
I convert this to a discussion since this doesn't describe a singular issue. There might be several bugs/issues in what you describe that are worth opening new issues. But then there needs to be a description of the issue, how to reproduce it, basic information of the system (OS, etc.). I know, we should have a template for this…

**Reply by OConnorStP** [reply]
Hi Nils.  Thanks for tidying that up.

I think the cause of the error is that the button up on top of Aaron's thread about spinning ball of death invited me to click it to create a new Issue, so I presumed that I was creating a new thing like the last one.   I don't think I know enough about Github to be aware of a template if you created it.  So maybe what you did is better?  An experienced person transforms the new thing into the correct kind of new thing?

**Reply by ntonnaett** [reply]
No, we didn't create a template yet. It's just on my internal todo list. Right now people don't get any guidance when opening an issue what information developers need to work on it. If you only have a vague idea of an issue, opening a discussion is probably the better way. So we can distill the relevant information together for opening an issue.

Some questions I have: What operating system are you using? With what versions of JackTrip and JACK do you experience your problems? What do you mean with one-way connections? What issues do you have with QJackCtl? Could you give us the log messages? Permission problems sound like you are using macOS. AFAIK users should get a dialog to grant all necessary permissions. Is this not working for you?

I thought that the JackTrip installer on Windows bundled JACK at some point in time, but because the RtAudio backend is sufficient for most use cases, we don't do this anymore?! I don't use Windows and didn't follow the development there.

**Comment by OConnorStP** [comment]
Hi Aaron,

Sorry about the delay.  I've been fighting Jacktrip issues a lot this week and kinda wanted to see if I could see any patterns on the path to replicating some of this stuff.  Alas, it's thin soup.  Y'know what a cool thing would be?  A gizmo like this test that we use for diagnosing video issues (we are using VDO.ninja a lot these days -- you should check it out if you haven't come across it).  Here's a link to the test.  It gives you a really good feel as to whether the person is going to be able to sustain a solid video connection.

https://vdo.ninja/speedtest

Something like this for Jacktrip would be an immense help.  It could scurry around and look for Jack, do a little performance testing (connectivity, bandwidth, jitter, etc.), make sure all the ports are working, etc. etc.  and give people little hints about what it finds.  My kingdom for such a thing.

Back to this thread.  I've been on long Jacktrip sessions for the last several days and each one needed to be force-quit.  My setup is a headless M1 Monterey client running 1.7.1 talking to a Debian 12 server running 1.6.7 or 1.4.1.  Short sessions work fine.  In longer sessions I come back to the headless M1 to find the GUI locked up but the Jacktrip connection running fine.  

I was working this morning with a performer and her copy of 1.7.1 locked up much more quickly but we were sorting out lots of weird stuff with Jacktrip and I can't recall the steps we'd taken before it froze.  I noticed that she immediately force-quit the app and when I commented about that she said that it happens all the time so she just does it automatically now.   This from a decidedly non-technical user.

Back to the Jack stuff (different from the GUI stuff).  Working with another performer who's connection has been really weird, i noticed that she has two pretty-fresh installations of Jacktrip on two different Intel Macs.  Doing Jacktrip -v from the command line on one computer gives the "WEAK-Jack:  OK. (0)" reply.  The other one comes back with "WEAK-JACK: libjack was not found".   Again this was in the middle of a long trail of troubleshooting and just something I noted at the time.  Since presumably these are Jacktrip-generated messages, it would be really helpful to expand the libjack one to say a little bit more about what is implied.  Is it a problem?  if so what?  and what should be done?

Maybe the little diagnostic toolkit I'm dreaming of could be built into the command-line version?  Maybe behind the -v parm?

Along the same lines, maybe the error message that says something like "this port is already binded" could be reworded to give similar hints about what that implies.  My reaction to that message is "everything is borked, shut down and restart your computer I'll see you on the other side" but that's not implied by the message, it's just my personal reaction.  An aside, I've always wondered where the non-word "binded" came from and why it hasn't been changed over all these years.

Since Jacktrip doesn't install Jack, let me rephrase the question at the top of the thread.  Have there been changes to the way Jacktrip decides when and how to launch Jack?  Maybe that's what's triggering some of this.  

Here's another anecdote.  A couple of experienced Jacktrip performers, preparing for a conference performance, were stuck.  All kinds of weird connection failures.  We dinked around together for a few hours and discovered that if we paid careful attention to align the way that command-line Jacktrip and GUI Jacktrip were configured when they were (successfully) shut down, we could get things working.  But it seems like getting those two out of sync caused a lot of the troubles they were seeing.  And, since Jacktrip being force-quit a lot, those things tend to crop up a lot.  Again, no way I can reproduce the details of that hour-long session in order to reproduce it.

I have several other examples of equally odd behavior, but you get the picture.  Here's another wishlist item -- could we get a "reference" version of Jacktrip that is totally solid in Classic mode and then let users elect to stay on that one and not be catapulted into the newest version when they start up the next time.  In the case of my largest performance group (20-25 people) we've gotten in the habit of spending 2 hours in advance of rehearsal/performance sessions just getting people logged into Jacktrip -- because almost every time there's some new wrinkle to work out.  

thus ends...  the tirade...  of a really tired Jacktrip-support guy.  :-(

**Reply by psiborg112** [reply]
Unfortunately a test app would be really outside of the scope of what we're doing here, and I don't think anyone on the dev team (at least the open source side) has the time to dedicate to a project like that. (Also there are plenty of existing jitter and bandwidth tests online to check your connection.)

So, to quickly go over your questions: I haven't been able to reproduce the lock up issue, although to be fair, most of my long running sessions have been on linux. I'm planning to leave a mac running jacktrip for the next few hours to see if I can finally reproduce it. As mentioned on the issues page though, I've replaced a significant portion of the meter code recently which may or may not have some bearing on this issue. (And that'll be out in 1.8.0.)

The WEAK-Jack message is a very technical one that isn't created by JackTrip but by one of the libraries it relies on. If we didn't use it the program wouldn't run without an install of JACK. libjack not found means precisely what it sounds like: JACK isn't installed (or at least not properly installed) on the computer being used. Originally in the GUI there was an error message to accompany this to let the user know that the program was falling back to the RtAudio backend (unless the user had explicitly chosen RtAudio previously). With more people using RtAudio and not bothering with JACK, others were worried that the warning would scare users. I argued to keep a reworded version of the warning, but it didn't get up, so it only exists in my builds. So if you see libjack not found, you'll be using RtAudio. It's worth checking the Audio Backend tab to see exactly what is going on.

This port is already bound (yes it should be fixed to that) just means that something else on your computer is already using the network port that JACK wants to use. Closing anything that is currently using the port (which is why a reboot works) or changing the local port setting (under the advanced options) will fix the issue. (And if you change the local port it means you'll need to change the remote port on the other side in P2P mode.)

And there's really no way to create a stable reference version. We don't plan for bugs to be in the code, but they inevitably will be. It's the nature of programming. Especially when there are ones that appear out in the wild that don't appear in our testing. (Like the lock up issue that I really haven't yet been able to recreate.) If you tie people to a 'reference' version, then they get to keep all of the bugs from that version. There's just no way around that. If you want the program to keep getting better, then freezing it in place, warts and all, isn't really the way to achieve that.

**Reply by OConnorStP** [reply]
Thanks Aaron, terrific response.  

I won't belabor the test-app wish.  I'm just struck by the customer-facing idea behind it.  It's simple and it immediately highlights a setup that isn't going to work very well.  I've been using it a lot when diagnosing bad Jacktrip connections and it'll work for me.  Maybe just add a reference to it, and an explanation of what it's identifying, in the 'having trouble with Jacktrip?' documentation and call it done?  

I've been watching all your work on the meter code and look forward to trying it out.  

Regarding your 'graph that leads off with the WEAK-jack message -- this is really at the heart of the issue.  Once upon a time users perceived Jacktrip and Jack as totally different things and were taught how to load Jack and then Jacktrip.  With newer code, that relationship has become blurry and confusing (witness my confusion about whether Jacktrip installs Jack).  I like the approach you took with the error messages because it reduces that ambiguity.  

I'm lobbying for more error-message work to get accurate help to stuck-users at the time that they're stuck.  The documentation about installing Jacktrip (I'm looking [HERE](https://help.jacktrip.org/hc/en-us/articles/360053942493-Setup-Guides)) seems out of date, given all the changes that have happened since those documents were written.  Should people be advised to install Jack?  Should they uninstall Jack if they have it?  How do they do that?  If you are seeing libjack-not-found, what choices do you have, what should you do?  etc. etc.  

Once upon a time, I'd have lobbied that we rally round the documentation problem.  I'm spent.  Now I'm lobbying that the people who are closest to the issues (you folks who write the code) try to get help to people through reworked error-messages.

On to port-already-bound (yep, I agree).  I think you wrote a pretty good first draft of the slightly-longer version of that error message in that paragraph.  Maybe something like "This port is already bound by another process that is already running, most likely another instance of Jacktrip that didn't close properly..." followed by a brief suggestion about what people should do. 

And to your last 'graph.  I know.  No pain no gain.  There's just been a lot of pain.

**Reply by psiborg112** [reply]
I think part of the reason that the relationship has become blurry is because JACK isn't needed for JackTrip anymore with the RtAudio backend. And in the official builds (unlike in my builds) there's never any notification to the user that JACK isn't there. (Beyond the cryptic WEAK-Jack message.) And for a lot of users (particularly those using VirtualStudio mode) who don't care about the underlying technicalities, that's not an issue. The problem is when you have people who actually want to do more with the app, who start to learn about JACK unaware that they're not actually set up to use it.

It comes down to different software design paradigms, and how these deal with the problems that invariably arise with complex applications. Do we insulate the user from knowing anything about the underlying backend and automatically make choices that we think are in their best interests? Or do we try to inform them about what's going on and hope that they'll be able to navigate their way through the technicalities? (I definitely prefer the latter approach. And now that the VS crew have their own interface, it may be possible for us to push things in this direction a little more going forward.)

There's also really no one right decision about whether to install JACK or not. I personally always use it, because it's the foundation of my entire audio setup on linux. And because the ability to route audio to and from anything is something I find particularly important and useful. (And it's why I also teach all of my students to use JACK rather than relying on RtAudio.) For someone who just wants to send directly from their interface, it's probably not so crucial.

**Reply by psiborg112** [reply]
Aha! I found it! The discussion on the RtAudio warning: #477

**Reply by psiborg112** [reply]
"I think the way that we're probably headed here is that the warning will remain in my build with the new wording that I've proposed and won't exist in the official build, but I think in the process we could well be making life more difficult for our more intermediate and advanced users."

**Reply by OConnorStP** [reply]
Aha indeed!   I love error messages like those!   

I enthusiastically support your "inform them about what's going on and hope that they'll be able to navigate their way through the technicalities" approach.  I would slightly reword it to say "inform them about what's going on in a way that will help them navigate their way through the technicalities."  I used to hope for documentation to do this, but think error messages are a better approach right now.  Deliver the tips when they're needed most.

I'd love to put a lot of time into being on the error-message team.  :-)

**Reply by OConnorStP** [reply]
Ah.  Our messages crossed.  Are your builds still a thing?  Can I switch over to them with my gang(s)?

**Reply by psiborg112** [reply]
Yes, you're more than welcome to. They don't have any of the VS stuff, so they're only usable in classic mode. You'll find them [here](https://www.psi-borg.org/other-dev.html). They also already include the new meter code. (Which has now been running for a little over 2.5 hours on my mac without issue, so here's hoping.)

**Reply by OConnorStP** [reply]
Great.  I've added your builds front and center on [www.JCKTRP.com](https://jcktrp.com/) and have a copy on the machine that hangs.  I've got a couple long sessions coming up in a couple days and will report back.

Thanks!

**Reply by psiborg112** [reply]
Thanks. Let me know how it goes. If it solves the issue that means it'll be fixed in the next release of the official builds as well.
